gluster interface for gd1 and gd2

vredara opened the pull request:
added interface to run specific cmds against gd1 and gd2
gd1.go has gd1 related functions, gd2 has glusterd2 related functions
glusterutils.go has interface which gd1 and gd2 has to implement

Signed-off-by: venkata edara <redara@redhat.com>

shtripat commented on pkg/glusterutils/glusterutils.go:
```diff
@@ -0,0 +1,39 @@
+package glusterutils
```
I feel this is candidate for `glusterutils/exported.go`

vredara commented:
I merged the vol info into respective interfaces. I also added MakeGluster func to create GD1 or GD2 based on config. all gd1 related can go to gd1.go and gd2 related funcs can be in gd2.go.

vredara commented:
rebased it with master and did commit again.

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -13,25 +14,14 @@ var (
-	setDefaultConfig(config)
-	peerList, err := Peers(config)
+func (g GD1) IsLeader(config *Config) (bool, error) {
+	setDefaultConfig(g.config)
```
`setDefaultConfig` can be moved to `MakeGluster` since it is required by all calls

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -13,25 +14,14 @@ var (
-func IsLeader(config *Config) (bool, error) {
-	setDefaultConfig(config)
-	peerList, err := Peers(config)
+func (g GD1) IsLeader(config *Config) (bool, error) {
```
- config parameter can be removed since already available as `g.config`
- change `(g GD1)` to `(g *GD1)`

aravindavk replied:
We also need `func (g *GD2) IsLeader() (bool, error)`

```
func (g *GD1) IsLeader() (bool, error) {
	peerList, err := Peers(config)
	if err != nil {
		return false, err
	}
	peerID, err := LocalPeerID(config)
	if err != nil {
		return false, err
	}

	var maxPeerID string
	//This for loop iterates among all the peers and finds the peer with the maximum UUID (lexicographically)
	for i, pr := range peerList {
		if pr.Online {
			if peerList[i].ID > maxPeerID {
				maxPeerID = peerList[i].ID
			}
		}
	}
	//Checks and returns true if maximum peerID is equal to the local peerID
	if maxPeerID == peerID {
		return true, nil
	}
	return false, nil
}

func (g *GD2) IsLeader() (bool, error) {
	peerList, err := Peers(config)
	if err != nil {
		return false, err
	}
	peerID, err := LocalPeerID(config)
	if err != nil {
		return false, err
	}
	//The following lines checks and returns true if the local PeerID is equal to that of the first PeerID in the list
	for _, pr := range peerList {
		if pr.Online {
			if peerID == peerList[0].ID {
				return true, nil
			}
			return false, nil
		}
	}
	// This would imply none of the peers are online and return false
	return false, nil
}
```

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -13,25 +14,14 @@ var (
 		return false, err
 	}
-	peerID, err := LocalPeerID(config)
+	peerID, err := LocalPeerID(g.config)
```
Also add `LocalPeerID` to interface as mentioned earlier. and use `peerID, err := g.LocalPeerID()`

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -22,6 +23,15 @@ func VolumeInfo(config *Config) ([]Volume, error) {
+*/
+
+func MakeGluster(config *Config) ginterface {
+	if config.GlusterMgmt == "" || config.GlusterMgmt == "glusterd" {
```
`if config.GlusterMgmt == MgmtGlusterd`

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -64,15 +54,12 @@ func IsLeader(config *Config) (bool, error) {
-		return gd2VolumeInfo(config)
+func MakeGluster(config *Config) ginterface {
+	if config.GlusterMgmt == "" || config.GlusterMgmt == "glusterd" {
+		return GD1{config: config}
```
Also call `setDefaultConfig(config)`

```
setDefaultConfig(config)
if config.GlusterMgmt == MgmtGlusterd {
    return GD1{config: config}
}
return GD2{config: config}
```

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -56,3 +66,44 @@ func LocalPeerID(config *Config) (string, error) {
+}
+
+// GetGlusterVersion gets the glusterfs version
+func GetGlusterVersion() (string, error) {
```
this comment not addressed

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -114,3 +101,24 @@ func LocalPeerID(config *Config) (string, error) {
+}
+
+// ExecuteCmd enables to execute system cmds and returns stdout, err
+func ExecuteCmd(cmd string) ([]byte, error) {
```
exposing this function not required. change name to `executeCmd`

aravindavk commented on pkg/glusterutils/types.go:
```diff
@@ -88,3 +88,28 @@ type Volume struct {
+type ginterface interface {
+	Peers() ([]Peer, error)
+	HealInfo(vol string) ([]HealEntry, error)
+	VolumeInfo(vol string) ([]Volume, error)
```
add `IsLeader` and `LocalPeerID`

aravindavk commented on pkg/glusterutils/healinfo_gd1.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+// GetHealInfo gets gluster vol heal in []HealEntries
+func (g GD1) HealInfo(vol string) ([]HealEntry, error) {
```
change `(g GD1)` to `(g *GD1)`

aravindavk commented on pkg/glusterutils/healinfo_gd2.go:
```diff
@@ -0,0 +1,21 @@
+package glusterutils
+
+// GetHealInfo gets heal info from glusterd2 using rest api
+func (g GD2) HealInfo(vol string) ([]HealEntry, error) {
```
change `(g GD2)` to `(g *GD2)`

aravindavk commented on pkg/glusterutils/peers_gd1.go:
```diff
@@ -19,10 +19,10 @@ type peersGlusterd1 struct {
 
 // PeersGD1 returns the list of peers ( for GlusterD1 )
-func peersGD1(config *Config) ([]Peer, error) {
+func (g GD1) Peers() ([]Peer, error) {
```
change `(g GD1)` to `(g *GD1)`

aravindavk commented on pkg/glusterutils/peers_gd2.go:
```diff
@@ -9,9 +9,9 @@ var (
 
 // PeersGD2 returns the list of peers ( for GlusterD2 )
-func peersGD2(config *Config) ([]Peer, error) {
+func (g GD2) Peers() ([]Peer, error) {
```
change `(g GD2)` to `(g *GD2)`

aravindavk commented on pkg/glusterutils/volinfo_gd1.go:
```diff
@@ -7,77 +7,9 @@ import (
-}
-
-func gd1VolumeInfo(config *Config) ([]Volume, error) {
+func (g GD1) VolumeInfo(volname string) ([]Volume, error) {
```
change `(g GD1)` to `(g *GD1)`

aravindavk commented on pkg/glusterutils/volinfo_gd2.go:
```diff
@@ -11,12 +11,12 @@ var subvolTypeValueToName = map[api.SubvolType]string{
 
-func gd2VolumeInfo(config *Config) ([]Volume, error) {
-	client, err := initRESTClient(config)
+func (g GD2) VolumeInfo(volname string) ([]Volume, error) {
```
change `(g GD2)` to `(g *GD2)`

aravindavk commented on pkg/glusterutils/volinfo_gd1.go:
```diff
@@ -7,77 +7,9 @@ import (
+func (g GD1) VolumeInfo(volname string) ([]Volume, error) {
 	// Run Gluster volume info --xml
-	out, err := exec.Command(config.GlusterCmd, "volume", "info", "--xml").Output()
+	out, err := exec.Command(g.config.GlusterCmd, "volume", "info", volname, "--xml").Output()
```
empty string in argument may be problem try running `gluster volume info "" --xml`

shtripat commented on pkg/glusterutils/healinfo_gd1.go:
```diff
@@ -0,0 +1,33 @@
+package glusterutils
```
I would suggest heal info to be done in another PR and not mix in this.
Lets take the interface changes in one PR first and then send another PR for heal info.

vredara replied:
ack

aravindavk commented on gluster-exporter/metric_brick.go:
```diff
@@ -129,14 +129,15 @@ func diskUsage(path string) (disk DiskStatus, err error) {
 
 func brickUtilization() {
-	volumes, err := glusterutils.VolumeInfo(&glusterConfig)
+	gluster := glusterutils.MakeGluster(&glusterConfig)
```
This can be moved to `main.go`. Make `gluster` as global variable similar to `glusterConfig`

vredara replied:
the actual config check is done in main() function in main.go. if i make gluster = MakeGluster(&config) before we parse the global conf, doesnt it always return gd1 instance?.

aravindavk replied:
it should be after parsing the config and before starting the metrics collectors

vredara replied:
I just declared gluster var as global and assigned it in main().

aravindavk commented on gluster-exporter/metric_ps.go:
```diff
@@ -71,6 +71,8 @@ var (
 		labels,
 	)
+
+	gluster = glusterutils.MakeGluster(&glusterConfig)
```
same as above

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -63,28 +40,73 @@ func IsLeader(config *Config) (bool, error) {
+	if err != nil {
+		return "", err
 	}
+	defer func() {
```
below code is duplicated. create one more internal function and call from both `GD1.LocalPeerID` and `GD2.LocalPeerID`

vredara replied:
ack

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -114,3 +136,24 @@ func LocalPeerID(config *Config) (string, error) {
+	return fields[1], err
+}
+
+// ExecuteCmd enables to execute system cmds and returns stdout, err
```
nit: code comment reflects old name

aravindavk commented on pkg/glusterutils/volinfo_gd1.go:
```diff
@@ -7,77 +7,9 @@ import (
-}
-
-func gd1VolumeInfo(config *Config) ([]Volume, error) {
+func (g GD1) VolumeInfo() ([]Volume, error) {
```
not changed to `(g *GD1)`

aravindavk commented on pkg/glusterutils/volinfo_gd2.go:
```diff
@@ -11,8 +11,8 @@ var subvolTypeValueToName = map[api.SubvolType]string{
 
-func gd2VolumeInfo(config *Config) ([]Volume, error) {
-	client, err := initRESTClient(config)
+func (g GD2) VolumeInfo() ([]Volume, error) {
```
same here, not changed to `(g *GD2)`

aravindavk commented:
also please squash all commits into one.

vredara commented:
Incorporated all comments.

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
+	if err != nil {
+		return "", err
+	}
+	peerid, err := readPeerID(fileStream, keywordID)
```
directly return `return readPeerID(fileStream, keywordID)`

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
+	}
+	peerid, err := readPeerID(fileStream, keywordID)
+	return peerid, err
+	/*
```
please remove commented code

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
+	if err != nil {
+		return "", err
+	}
+	peerid, err := readPeerID(fileStream, keywordID)
```
`return readPeerID(fileStream, keywordID)`

aravindavk commented on pkg/glusterutils/exporterd.go:
```diff
@@ -114,3 +98,101 @@ func LocalPeerID(config *Config) (string, error) {
+	peerid, err := readPeerID(fileStream, keywordID)
+	return peerid, err
+
+	/*
```
remove commented code

aravindavk commented on gluster-exporter/main.go:
```diff
@@ -28,6 +28,7 @@ var (
 	config                        = flag.String("config", defaultConfFile, "Config file path")
 	defaultInterval time.Duration = 5
 	glusterConfig   glusterutils.Config
+	gluster         glusterutils.ginterface
```
Unexported name used here

vredara commented:
I could not see metric_volume_counts.go in my git env. I rebased the branch and fixed the build issues. turned ginterface to GInterface , now the name is exposed and can be used in main.go

aravindavk commented:
Please check the build failure.

vredara commented:
> Please check the build failure.

metric_volume_counts.go is not visible to me, so I  am not able to change code. I fetched the latest master , rebased but still metric_volume_counts.go is not there. will work on this.

aravindavk commented:
```
git checkout master
git fetch upstream
git merge upstream/master
git checkout gluster-utils
git rebase -i master
git push origin gluster-utils -f
```

aravindavk commented:
@vredara thanks for the PR
